spi/src/main/java/org/jboss/pnc/spi/datastore/predicates/BuildRecordPredicates.java

```diff
@@ -382,7 +382,10 @@ public static Predicate<BuildRecord> withoutImplicitDependants() {
                             // All entries in a column have to be NULL which is equivalent to count being 0 (aggregation
                             // functions ignore NULLs)
                             cb.equal(cb.count(buildRecordImplicitDependants.get(BuildRecord_.id)), 0)));
-            return cb.and();
+            return cb.or(
+                    // build can depend on itself
+                    cb.notEqual(root.get(BuildRecord_.id), buildRecordImplicitDependants.get(BuildRecord_.id)),
+                    cb.isNull(buildRecordImplicitDependants.get(BuildRecord_.id)));
```

Comment:
what was the reason for adding this isNull condition?

This query needs non-null `buildRecordImplicitDependants.get(BuildRecord_.id)` so that the having clause can exclude builds by built-artifacts that have a dependant. This part effectively deletes all built artifacts that are used as a dependency by other build and makes the count always 0 even for builds that have a dependant artifact.

Reply:
because criteria api does not have "is distinct from" (nullsafe `!=`). We need to keep the null because otherwise the lines will completly disappear and it will return nothing. Without the isNull it will keep only lines with non-null `buildRecordImplicitDependants.get(BuildRecord_.id)`, so count would be alwas > 0 and nothing would be ever returned.
Also note that it's or.

Reply:
gotcha